docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,461 @@
+The resources will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#NetworkQoS_TABLE).
+The `NetworkQoS` also has `status` field which is populated by ovn-k which helps users to identify whether NetworkQoS rules are configured correctly in OVN or not.
+
+ovn-controller has this configuration, ovn-encap-tos:
+
+-------8<-------------8<--------
+
+external_ids:ovn-encap-tos
+
+ovn-encap-tos indicates the value to be applied to OVN tunnel interface’s option:tos as specified in the Open_vSwitch database Interface table. Please refer to Open VSwitch Manual for details.
+
+   options : tos: optional string
+          Optional. The value of the ToS bits to be set on the encapsulat‐
+          ing packet. ToS is interpreted as DSCP and ECN  bits,  ECN  part
+          must be zero. It may also be the word inherit, in which case the
+          ToS will be copied from the inner packet if it is IPv4  or  IPv6
+          (otherwise  it  will be 0). The ECN fields are always inherited.
+          Default is 0.
+
+-------8<-------------8<--------
+
+which when set to inherit will take the marking from the inner header and apply it to the GENEVE packet. Therefore, east west traffic within the cluster can benefit
+from this API and not just the north south traffic.
+
+### API Details
+
+* A new API `NetworkQoS` under the `k8s.ovn.org/v1` version will be added to `go-controller/pkg/crd/networkqos/v1`.
+This would be a namespace-scoped CRD:
+
+```go
+
+import (
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=networkqoses
+// +kubebuilder::singular=networkqos
+// +kubebuilder:object:root=true
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=".status.status"
+// +kubebuilder:subresource:status
+// NetworkQoS is a CRD that allows the user to define a DSCP marking and metering
+// for pods ingress/egress traffic on its namespace to specified CIDRs,
+// protocol and port. Traffic belong these pods will be checked against
+// each Rule in the namespace's NetworkQoS, and if there is a match the traffic
+// is marked with relevant DSCP value and enforcing specified policing/shaping
+// parameters.
+type NetworkQoS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   Spec   `json:"spec,omitempty"`
+	Status Status `json:"status,omitempty"`
+}
+
+// Spec defines the desired state of NetworkQoS
+type Spec struct {
+	// networkAttachmentName selects the network-attachment-definition for
+	// which the QoS Rules will be applied. The NAD can be of any type Layer-3,
+	// Layer-2 or Localnet. If not specified, the default cluster-wide network
+	// will be used.
+	// +optional
+	NetworkAttachmentName string `json:"networkAttachmentName,omitempty"`
```

Comment:
should we rethink this with UDNs..? maybe we should use networkName that the API should be part of instead of the NADName? calling this NetAttchDefName makes it tied to NADs only

Reply:
@pperiyasamy and @tssurya : We have made mods to this Enhancement, so it can support UDN as well as NAD.
Please see: https://github.com/pperiyasamy/ovn-kubernetes/pull/1
In summary, by replacing the NetworkAttachmentName with a ref, we can use any `kind` we decide to support.
This change is reflected in the implementation we are interested in merging via https://github.com/ovn-org/ovn-kubernetes/pull/4508 .

Reply:
Thanks @flavio-fernandes ! sorry for the delay. @tssurya NetworkQoS CRD is currently a namespace scoped object. If we need to support NetworkQoS for UDNs, then i think we must extend it as a cluster scoped object and provide a subject in the spec  (as similar to [ANP](https://github.com/pperiyasamy/ovn-kubernetes/blob/master/go-controller/vendor/sigs.k8s.io/network-policy-api/apis/v1alpha1/adminnetworkpolicy_types.go#L76)) so that it can be applied on pods across namespaces served by same user defined network. Is that correct @tssurya ? I think we may still need to retain NetworkAttachmentName in the subject to support QoS on the secondary networks.

Reply:
we should check if two secondary UDNs can be attached to the same pod with both these UDNs having the same network name -> basically what we have for NADs do we have the same support for UDN CR?

Reply:
yes @tssurya, if we decide it be a namespaced scope object, then there is no need for a special handling in CRD, The associated UDN network can be found from namespace and can be handled accordingly (default network vs primary UDN) in implementation.